### PR TITLE
[RFR] [Rework] Datagrid rowclick toggle

### PR DESCRIPTION
**Pull request by fargito**
This is a rework of https://github.com/marmelab/react-admin/pull/3120

* Allow `rowClick="toggleSelection"` option on DatagridRow
* Add corresponding documentation
* Add tests for all the `rowClick` options

Closes https://github.com/marmelab/react-admin/pull/3120

### CODE REVIEW

**Comment by djhi** [comment]
You're on :fire: ! Thanks!

**Comment by djhi** [comment]
However, this is technically a new feature and should target the `next` branch which will be the `3.0`

**Comment by fargito** [comment]
Oops wait how do I only keep only my commits ?

File: packages/ra-ui-materialui/src/list/DatagridRow.js [hunk 1]
```diff
@@ -62,33 +62,36 @@ class DatagridRow extends Component {
...
+                push(linkToRecord(basePath, id, 'show'));
+                return;
+            case 'expand':
+                this.handleToggleExpanded(event);
+                return;
+            case 'toggleSelection':
+                this.handleToggle(event);
+                return;
+            default:
+                push(path);
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/list/DatagridRow.js]: If path is empty, the previous code didn't push. You're introducing  breaking change here.

File: packages/ra-ui-materialui/src/list/DatagridRow.spec.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+describe('<DatagridRow />', () => {
+    const defaultProps = {
+        id: 15,
+        basePath: '/blob',
+    };
+
+    const event = { preventDefault: () => {}, stopPropagation: () => {} };
+
+    describe('on click event, depending on rowClick prop', () => {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/list/DatagridRow.spec.js]:
```suggestion
    describe('rowClick', () => {
```

File: packages/ra-ui-materialui/src/list/DatagridRow.spec.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+                    {...defaultProps}
+                    rowClick={customRowClick}
+                    push={push}
+                />
+            );
+
+            await wrapper.instance().handleClick(event);
+            expect(push.mock.calls).toEqual([['/bar/foo']]);
+        });
+    });
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/list/DatagridRow.spec.js]: missing a test with a falsy value - it should do nothing

File: packages/ra-ui-materialui/src/list/DatagridRow.js (hunk 1 above)
Comment (fargito) [reply packages/ra-ui-materialui/src/list/DatagridRow.js]: No because there is a condition just before the `switch` :)

Comment (fargito) [reply packages/ra-ui-materialui/src/list/DatagridRow.js]: However I agree that it would be clearer with this condition in the `default`

**Comment by fzaninotto** [comment]
Thanks!

**Pull request merged**